lib/next_ls/diagnostic_cache.ex

```diff
@@ -1,33 +1,69 @@
 defmodule NextLS.DiagnosticCache do
-  # TODO: this should be an ETS table
   @moduledoc false
-  use Agent
+  use GenServer
 
   def start_link(opts) do
-    Agent.start_link(fn -> Map.new() end, Keyword.take(opts, [:name]))
+    opts = Keyword.take(opts, [:name])
+    GenServer.start_link(__MODULE__, opts, opts)
+  end
+
+  def init(opts) do
+    name = opts[:name]
+
+    ets =
+      if name do
+        :ets.new(name, [:named_table, :protected, :bag, {:read_concurrency, true}])
+      else
+        :ets.new(:diagnostic_cache, [:protected, :bag, {:read_concurrency, true}])
+      end
+
+    {:ok, %{ets: ets}}
   end
 
   def get(cache) do
-    Agent.get(cache, & &1)
+    cache
+    |> GenServer.call(:all)
+    |> to_map()
```

Comment:
This is purely done to preserve the old interface of the cache, I originally refactored all the places that use it but decided to preserve it as is. Let me know what you think would be better, I think using the new list structure directly at some places might be a performance benefit other than that the old nested map structure is more convenient for use.

There was also the approach of storing it in ets as maps but it turned out to be a hassle since there's no update like interface like `Map.update`